src/Stackage/Package/Hashes.hs

```diff
@@ -1,93 +1,148 @@
+import qualified Data.Conduit.List as CL
 import qualified Data.Map as Map
-import Distribution.Package (PackageName)
+import qualified Data.Set as Set
+import Data.Version (Version)
+import Distribution.Package (PackageName(..))
 import Network.HTTP.Client.Conduit
        (HttpException(StatusCodeException), parseRequest,
         responseCookieJar, responseHeaders, responseStatus)
-import Network.HTTP.Types (statusCode)
 import Network.HTTP.Simple (httpSink)
+import Network.HTTP.Types (statusCode)
 import System.FilePath (dropExtension)
 
 import Stackage.Package.Git
-import Stackage.Package.Locations
 import Stackage.Package.IndexConduit
+import Stackage.Package.Locations
+
 
 -- | Compares hashes in 'package.json' to the ones in the repo. In case a new
 -- package.json appears without coresponding hashes in the repo, a package is
 -- downloaded, hashes are computed and compared to ones in 'package.json' file.
--- 
-entryUpdateHashes
-  :: (MonadMask m, MonadIO m)
-  => GitRepository -> IndexEntry -> m ()
-entryUpdateHashes hashesRepo (PackageEntry IndexFile {ifFile = HackagePackage {..}
-                                                     ,..}) = do
-  mpackage <- createHashesIfMissing hashesRepo ifPackageName hackageVersion
-  case mpackage of
-    Nothing -> return ()
-    Just package ->
-      forM_ [tshow MD5, tshow SHA256] $
-      \hashType ->
-         unless
-           (Map.lookup (toLower hashType) (hHashes hackageHashes) ==
-            Map.lookup hashType (packageHashes package))
-           (error $
-            "Stackage.Hackage.Hashes.entryUpdateHashes: Hash " ++
-            unpack hashType ++
-            "value mismatch for: '" ++
-            getPackageFullName ifPackageName hackageVersion ++
-            "' computed vs one from Hackage.")
-entryUpdateHashes _ _ = return ()
+-- Returned is a map with all packages and their valid versions.
+sinkPackageHashes
+  :: (MonadIO m, MonadMask m)
+  => GitRepository
+  -> Consumer IndexEntry m (Map PackageName (Set Version))
+sinkPackageHashes hashesRepo = CL.foldM updateHashes Map.empty
+  where
+    updateHashes versionsMap (PackageEntry IndexFile { ifFile = HackagePackage {..}
+                                                     , ..
+                                                     }) = do
+      validHashes <-
+        createHashesIfMissing
+          hashesRepo
+          (hHashes hackageHashes)
+          ifPackageName
+          hackageVersion
+      return $
+        case validHashes of
+          False -> versionsMap
+          True ->
+            let with =
+                  Just .
+                  maybe
+                    (Set.singleton hackageVersion)
+                    (Set.insert hackageVersion)
+            in Map.alter with ifPackageName versionsMap
+    updateHashes versionsMap _ = return versionsMap
+
+
+-- | Generate package hashes file name.
+makeHashesFileName :: PackageName -> Version -> FilePath
```

Comment:
Is this for the upstream file location (inside 01-index.tar.gz) or in the Git repo? Now that we're dealing with JSON files with hashes in two different locations, it would be good to call it out explicitly.